### PR TITLE
 [KAFKA-7132] [WIP] Consider adding a faster form of rebalance 

### PR DESCRIPTION
**Pull request by ConcurrencyPractitioner**
Currently, when a consumer falls out of a consumer group, it will restart processing from the last checkpointed offset. However, this design could result in a lag which some users could not afford to let happen. For example, lets say a consumer crashed at offset 100, with the last checkpointed offset being at 70. When it recovers at a later offset (say, 120), it will be behind by an offset range of 50 (120 - 70). This is because the consumer restarted at 70, forcing it to reprocess old data. To avoid this from happening, one option would be to allow the current consumer to start processing not from the last checkpointed offset (which is 70 in the example), but from 120 where it recovers. Meanwhile, a new KafkaConsumer will be instantiated and start reading from offset 70 in concurrency with the old process, and will be terminated once it reaches 120. In this manner, a considerable amount of lag can be avoided, particularly since the old consumer could proceed as if nothing had happened. 

Here is the design doc for the pull request: 
https://cwiki.apache.org/confluence/display/KAFKA/KIP-333%3A+Add+faster+mode+of+rebalancing

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ConcurrencyPractitioner** [comment]
cc @becketqin @hachikuji

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 1]
```diff
@@ -77,6 +77,10 @@
     private final ConsumerInterceptors<?, ?> interceptors;
     private final boolean excludeInternalTopics;
     private final AtomicInteger pendingAsyncCommits;
+    private List<RebalanceConsumerCoordinator> rebalanceCoordinators = null;
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: RebalanceConsumerCoordinator -> ConsumerRebalancingCoordinator

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 2]
```diff
@@ -358,6 +373,17 @@ public boolean poll(final long timeoutMs) {
         return true;
     }
 
+    @Override
+    public boolean ensureActiveGroup(final long timeoutMs) {
+        if (useMultithreadRebalancing && rebalanceInProgress.get()) {
+            //fetch start offset
+            //fetch end offset (from which the consumer will start polling)
+            //create new RebalanceConsumer
+            rebalanceInProgress.compareAndSet(true, false);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Should the return value be checked ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 3]
```diff
@@ -163,6 +167,17 @@ public ConsumerCoordinator(LogContext logContext,
         addMetadataListener();
     }
 
+    public void setValue(final boolean useMultithreadRebalancing) {
+        if (this.useMultithreadRebalancing == null) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: What if this.useMultithreadRebalancing and useMultithreadRebalancing carry different values (one being true and the other being false) ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 4]
```diff
@@ -1031,6 +1033,7 @@ public void run() {
                         } else if (heartbeat.pollTimeoutExpired(now)) {
                             // the poll timeout has expired, which means that the foreground thread has stalled
                             // in between calls to poll(), so we explicitly leave the group.
+                            rebalanceInProgress.compareAndSet(false, true);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This flag is better set in LeaveGroupResponseHandler where we know whether the LeaveGroup request succeeded or not

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 3 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: The purpose of this setup is to make this.useMultithreadRebalancing effectively final. Once set to a non-null value, it can no longer be changed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 4 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: You have a point, I will make the change. :)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 2 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: I am not sure about the current design right now ... I will need to check.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 5]
```diff
@@ -1170,6 +1172,27 @@ void getStartAndEndOffsets() {
 
     }
 
+    private ConsumerRecords<K, V> checkRebalance(long timeoutMs) {
+        if (coordinator.isRebalancing() && useParallelRebalance) {
+            getStartAndEndOffsets();
+            if (rebalanceConsumer == null) {
+                rebalanceConsumer = new RebalanceKafkaConsumer(configs, startOffsets, endOffsets);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Where is RebalanceKafkaConsumer started ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumer.java [hunk 6]
```diff
@@ -75,14 +81,39 @@ public void sendRequest(Duration waitTime,
         this.callback = callback;
     }
 
-    public Set<TopicPartition> unfinishedPartitions() {
-        return unfinished;
+    private void alterSubscription(Set<TopicPartition> newSubscription) {
+        if (endOffsets.keySet().containsAll(newSubscription)) {
+            newSubscription.removeAll(endOffsets.keySet());
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumer.java]: Is this call equivalent to newSubscription.clear() ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumer.java [hunk 7]
```diff
@@ -75,14 +81,39 @@ public void sendRequest(Duration waitTime,
...
+        for (Map.Entry<TopicPartition, Long> entry : startOffsets.entrySet()) {
+            super.seek(entry.getKey(), entry.getValue());
+        }
+        this.endOffsets = endOffsets;
+        this.unfinished = startOffsets.keySet();
+    }
+
+    public void addNewOffsets(Map<TopicPartition, Long> startOffsets,
+                              Map<TopicPartition, Long> endOffsets) {
+        incomingStartOffsets.add(startOffsets);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumer.java]: incomingStartOffsets is a List where order matters.
startOffsets and endOffsets are Map's where traversal order may be non-deterministic.

For each key in startOffsets, maybe find corresponding key from endOffsets and add its value ?

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumerTest.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Test
+    public void testPositionsAfterInstantiation() {
+        final RebalanceKafkaConsumer<byte[], byte[]> consumer = newConsumer();
+        assertTrue(consumer.position(t1p) == 0L);
+        assertTrue(consumer.position(t2p) == 1L);
+    }
+
+    @Test
+    public void testConsumerIfContainsSubsciption() {
```
Comment (tedyu) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/RebalanceKafkaConsumerTest.java]: There is typo in Subscription

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 9]
```diff
@@ -1385,8 +1385,8 @@ public boolean matches(final AbstractRequest body) {
                 return body instanceof FetchRequest && ((FetchRequest) body).fetchData().containsKey(tp0);
             }
         }, fetchResponse(tp0, 1, 1), node);
-        time.sleep(heartbeatIntervalMs);
-        Thread.sleep(heartbeatIntervalMs);
+        time.sleep(heartbeatIntervalMs * 3);
```
Comment (tedyu) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Why is this change needed ?

Comment (ConcurrencyPractitioner) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Oh, I was tinkering with the test. Its not a neccesary change ... forgot to remove it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 10]
```diff
@@ -720,25 +733,17 @@ private KafkaConsumer(ConsumerConfig config,
...
                     config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
                     config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG),
                     config.getInt(ConsumerConfig.SEND_BUFFER_CONFIG),
                     config.getInt(ConsumerConfig.RECEIVE_BUFFER_CONFIG),
                     config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
-                    time,
-                    true,
-                    new ApiVersions(),
-                    throttleTimeSensor,
+                    time, true, new ApiVersions(), throttleTimeSensor,
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:745]: It's better to keep the parameters aligned (having same indentation)

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 11]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+            if (rebalanceConsumer == null) {
+                rebalanceConsumer = new RebalanceKafkaConsumer(configs,
+                                                               null,
+                                                               null,
+                                                               new HashMap<>(),
+                                                               new HashMap<>());
+                consumerThread = new Thread(rebalanceConsumer);
+            }
+            if (coordinator.isRebalancing(false)) {
+                System.out.println("Starting thread");
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Use log object if this is to be kept

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 12]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+
+    private void checkRebalance() {
+        if (useParallelRebalance) {
+            if (rebalanceConsumer == null) {
+                rebalanceConsumer = new RebalanceKafkaConsumer(configs,
+                                                               null,
+                                                               null,
+                                                               new HashMap<>(),
+                                                               new HashMap<>());
+                consumerThread = new Thread(rebalanceConsumer);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: consumerThread -> rebalancingConsumerThread

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 13]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+                                          RebalanceKafkaConsumer.ConsumerRequest.POLL,
+                             null,
+                                          new DefaultTaskCompletionCallback());
+        }
+    }
+
+    private RebalanceKafkaConsumer.RequestResult pollForResults(final long timeoutMs, final long now) {
+        long elapsed = time.milliseconds() - now;
+        boolean condition = remainingTimeAtLeastZero(timeoutMs, elapsed) != 0;
+        while (result == null && condition) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1223]: Since condition is just a comparison, you can put the comparison here directly

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 14]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+            try {
+                Thread.sleep(retryBackoffMs);
+            } catch (InterruptedException exc) { }
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Restore interrupt status by calling Thread.currentThread().interrupt();

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 15]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+            elapsed = time.milliseconds() - now;
+            condition = remainingTimeAtLeastZero(timeoutMs, elapsed) != 0;
+        }
+        return result;
+    }
+
+    private ConsumerRecords<K, V>  mergeRecords(final ConsumerRecords<K, V> records1, final ConsumerRecords<K, V> records2) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: This can be moved to ConsumerRecords class

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 16]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+        final HashMap<TopicPartition, List<ConsumerRecord<K, V>>> map = new HashMap<>();
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: nit: when records2 is empty, you can return immediately.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 17]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+            if (!map.containsKey(partition)) {
+                map.put(partition, records2.records(partition));
+            } else {
+                map.get(partition).addAll(records2.records(partition));
+            }
+        }
+        return new ConsumerRecords<>(map);
+    }
+
+    private ConsumerRecords<K, V> processRecords(final long timeoutMs,
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Add javadoc for the parameters

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 18]
```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
...
+        if (fetcher.sendFetches() > 0 || client.hasPendingRequests()) {
+            client.pollNoWakeup();
+        }
+
+        if (consumerThread != null && consumerThread.isAlive()) {
+            ConsumerRecords<K, V> offsetLagRecords =
+                    result == null ? (ConsumerRecords<K, V>) pollForResults(timeoutMs, checkRebalanceStart).value
+                            : (ConsumerRecords<K, V>) result.value;
+            if (offsetLagRecords == null) {
+                return this.interceptors.onConsume(new ConsumerRecords<>(records));
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1279]: nit: extract the call to this.interceptors.onConsume(new ConsumerRecords<>(records)) above line 1258 - its result would always be used

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 19]
```diff
@@ -1187,33 +1306,40 @@ public void assign(Collection<TopicPartition> partitions) {
...
         }
     }
 
+    /**
+     * Indicates whether or not the child consumer is alive.
+     *
+     * @return true if the secondary consumer thread created is alive.
+     *         false if not alive or has null value
+     */
+    public boolean childConsumerIsAlive() {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1348]: rebalanceConsumerIsAlive

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 20]
```diff
@@ -1480,10 +1618,46 @@ public void commitAsync(OffsetCommitCallback callback) {
...
+            if (!isChildConsumer && useParallelRebalance) {
+                checkRebalance();
+                final RebalanceKafkaConsumer.OffsetInclusion offsetInclusion = RebalanceKafkaConsumer.getRanges(rebalanceConsumer, offsets);
+                final HashMap<TopicPartition, OffsetAndMetadata> parentConsumerMetadata = offsetInclusion.getParentConsumerMetadata();
+                final HashMap<TopicPartition, OffsetAndMetadata> childConsumerMetadata = offsetInclusion.getChildConsumerMetadata();
+                final long hashCode1 = parentConsumerMetadata.hashCode();
+                final long hashCode2 = childConsumerMetadata.hashCode();
+
+                rebalanceConsumer.setOptionalInputArgument(childConsumerMetadata, hashCode1, hashCode2);
+                rebalanceConsumer.sendRequest(null,
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: What if only one of this call and the commitOffsetsAsync call on line 1642 succeeds ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 21]
```diff
@@ -2036,8 +2271,30 @@ public void resume(Collection<TopicPartition> partitions) {
...
         acquireAndEnsureOpen();
         try {
+            if (!isChildConsumer && useParallelRebalance) {
+                checkRebalance();
+
+                rebalanceConsumer.sendRequest(timeout, RebalanceKafkaConsumer.ConsumerRequest.END_OFFSETS, partitions, new DefaultTaskCompletionCallback());
+                final Map<TopicPartition, Long> offsets1 = fetcher.endOffsets(partitions, timeout.toMillis());
+                final Map<TopicPartition, Long> offsets2 = (Map<TopicPartition, Long>) pollForResults(timeout.toMillis(), time.milliseconds()).value;
+                final Map<TopicPartition, Long> result = new HashMap<>();
+                for (final TopicPartition partition : partitions) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: See if you can refactor this code which is similar to what beginningOffsets has (apart from the condition between pos1 and pos2)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 22]
```diff
@@ -59,7 +59,7 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
-import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.PriorityBlockingQueue;
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:63]: Why switch data structure ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 23]
```diff
@@ -163,6 +164,19 @@ public ConsumerCoordinator(LogContext logContext,
         addMetadataListener();
     }
 
+    // method will automatically set to false upon retrieving value
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: value is assigned instead of false, right ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 24]
```diff
@@ -583,19 +597,49 @@ public void close(final long timeoutMs) {
...
             OffsetCommitCompletion completion = completedOffsetCommits.poll();
             if (completion == null)
                 break;
-            completion.invoke();
+            final OffsetCommitCompletion previous = completions.isEmpty() ?
+                    new OffsetCommitCompletion(null, null, null, 0, 0) :
+                    completions.get(completions.size() - 1);
+            boolean containsMatch =
+                    !completions.isEmpty() && previous.hashCode == completion.hashCode;
+            if (completion.hashCode == 0) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: When would hashCode be 0 ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 25]
```diff
@@ -583,19 +597,49 @@ public void close(final long timeoutMs) {
...
+                completion.invoke();
+            } else if (containsMatch && previous.hashCodeEquals(completion)) {
+                completion.invoke(previous.offsets);
+            } else {
+                completions.add(completion);
+            }
+        }
+        for (OffsetCommitCompletion completion : completions) {
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Consider using java.util.Collections.addAll()

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 23 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Yeah, need to update.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 24 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: By default. When we are not splitting the offsets between the two consumers for committing, then we are setting the hashCode to zero.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 25 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: addAll() is not thread safe for PriorityBlockingQueue. Unlike add(). This way we could make it more thread safe.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 22 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:63]: Due to the current policy I have set up in the KIP, this part is adapted to accomadate for OffsetCommitCallback. Currently, we split offsets being committed into two, so now we have two CommitCallbacks. If both of them are called separately, then we have two calls to the callback. We cannot let that happen to ensure that Kafka's behavior remains the same. Therefore, we need both ```OffsetCommitCompletion```s in PriorityBlockingQueue before calling OffsetCommitCallback as one call instead of two. The Hashcodes are used to compare two OffsetCommitCompletions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 13 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1223]: Oh, I did this to make it shorter. (Wanted to make it more readable)

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 10 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:745]: NPath Complexity is a boon. Checkstyle believes KafkaConsumer's constructor contains too many lines. I did this to make the method shorter.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 14 above)
Comment (ConcurrencyPractitioner) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Oh, sure I will do that.
